Standardized sender signature (#30)
mobilecoinfoundation/mcips: text/0030-standardized-sender-sig.md

- Feature Name: Standardized Sender Signature
- Start Date: 2022-03-14
- MCIP PR: [mobilecoinfoundation/mcips#0030](https://github.com/mobilecoinfoundation/mcips/pull/0030)
- MobileCoin Epic: None

# Summary
[summary]: #summary

Enact MCIP #18 (Janus Mitigation), which sets the ephemeral key in the encrypted fog hint
to be the tx private key times the Ristretto base point. The original motivation was to enact
Janus attack mitigation.

We find a secondary use-case to motivate this: it becomes possible for the signer to create
a Schnorr signature over any payload, which a verifier can validate against the keys appearing
in a `TxOut`, without revealing the identity of the signer. We propose to both do #18 and standardize
the construction and validation of this signature.

# Motivation
[motivation]: #motivation

Often in a payments flow, it is necessary for the sender and recipient to exchange additional messages

> and metadata about the payment. Frequently, these messages need to be authenticated.

Reviewer: ```suggestion
and metadata about the payment, on or off-chain. Frequently, these messages need to be authenticated.
```

Reviewer 2: I feel like this doesn't make the motivation more clear -- i feel like the statement makes sense even in a non-blockchain context, and adding "on or off-chain" doesn't help

Reviewer: WFM!

We have a mechanism for doing this in the memo field, which attaches fixed-size payloads to TxOut's.
However, this imposes fairly strict size requirements on the data that can be sent.

Outside of the memo field, we have a notion of "receipts" and "confirmation numbers" which a sender
can use to identify themselves to a recipient. These are hashes of the TxOut shared secret, which
are similarly deniable, and convince the recipient that whoever sent them this number must have access
to secrets involved in the construction of the TxOut.

> However, these numbers are useless for convincing a third-party who is not a party to the transaction,

Reviewer: ```suggestion
However, these numbers are only capable of convincing a third-party who also has access to the recipients' private keys,
```

Or something like this - indicating that a 3rd party could be enlisted to do these confirmations, but only if they are provided with the correct material to do the validation.

Reviewer 2: fair!

Reviewer 2: i feel like, it is a bit weird to phrase things this way, because the recipient is not ever expected to give away their private keys, and if they are, that's an exceptional situation

Reviewer: Good point - and that's one of our motivating design constraints. The "numbers are useless for convincing a third-party" felt false in that if I really needed to convince the third party, I could provide them my private keys. Maybe rephrasing to say we don't want to put users in a situation where they need to provide their private keys to a third party is a design goal of MobileCoin?

Reviewer 2: thank you!

because only the recipient is able to validate these numbers. There is currently no form of receipt that
can be validated just from the blockchain.

This is desirable by some app project developers, who would like to store encrypted metadata per TxOut,
but off-chain. For example, they might like to have a database of encrypted metadata, keyed on TxOut's sent
by users of their app, and be able to ensure that only the true sender of the TxOut can post metadata associated
to the TxOut, without learning anything about the sender, recipient, amount, etc. of the TxOut.

After this proposal, a straightforward way to do this is that their app can sign the encrypted metadata using
the tx private key, and their server can check this signature against the public key in the encrypted fog hint.
We can simply use the Schnorrkel-based signature with Ristretto curve points that we already use in fog public addresses.

This might also be useful because it makes the notion of receipts somewhat more robust -- arbitrary metadata

> can be signed by the sender, and this form of signature will not have the deniability property.

Reviewer: Can you say more about why it does not have the deniability property?

Reviewer 2: The receipts based on the shared secret, have the deniability property, because both the sender and recipient can build the shared secret, and so form the signature. So it is technically possible for the recipient to have forged the signature, and for the sender to deny having created the receipt.

With this signature, the sender is signing it with the tx private key, which is a key only they have access to, and the recipient can't fake the signature using the view private key. So this signature is not deniable, it must have come from the sender.

Reviewer: Awesome - it may be worth including some of the text in your answer to define what we mean be deniability property

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

This MCIP will:

* Make it possible for the builder of a transaction output to additionally sign a blob
    using the tx private key.
* Make it possible for anyone to validate that signature against the TxOut, without
    knowing the identity of the sender.

# Reference-level explanation
[Reference-level-explanation]: #reference-level-explanation

We propose to

* Make internal changes to the transaction builder, following MCIP #18
  * Instead of being a second independent random private key, the private key used
    for `mc-crypto-box` encryption will be the same as the tx private key for that
    `TxOut`.
  * Also, implement code for a "Janus test" which enables clients to detect the
    Janus attack, as described in MCIP #18.
* Standardize a Schnorrkel-based signature using Ristretto curve points, made using the tx private key,
  and validated using the ephemeral key in the fog hint. This entails:

>   * Make the transaction builder either optionally return the tx private key when `add_output` is called,

Reviewer: Slight preference for signing at the time `add_output` is called, so that we don't have to build infrastructure around storing the private key for later use. Not sure whether we need to add anything to the MCIP about those tradeoffs, except maybe to consider the challenges of properly storing private key material in infrastructure.

Reviewer: The other side of the tradeoff is that if you want to send encrypted data about a previous TXO later, you may end up encrypting data for old TXOs when sending new TXOs, which could get messy to tease apart in the clients, or may not be so bad if we have a reasonable schema for this.

Reviewer 2: So what I had in mind was not that infrastructure stores the key, but that the app can create and submit a transaction, and then sign something perhaps a second or two later that goes to the app server.

It might be that the app needs to hit some endpoint to get some of the data that will be signed. If we allow that they can hold onto the ephemeral private key after the `TransactionBuilder` is gone then they might be able to pipeline their calls effectively. If not then they might be forced to serialize them.

It's just a thought experiment, I don't know if it would matter in a real app.

Reviewer: Ok cool - that makes sense to me

  off-chain data stores can all be fully independent.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

An alternative worth discussing is:

* Don't implement MCIP #18, don't make the ephemeral private key from the fog hint
  the same as the tx private key. Just make that private key available to form
  digital signatures with.

This meets the motivation of this MCIP, but since this takes us almost all the way
to MCIP #18, we think we should do MCIP #18 also, since it is valuable to mitigate
Janus attacks and that MCIP is well-studied. It costs us almost nothing to start
building transactions in the way described there, and at some point, flip the switch
on clients actually flagging Janus test failures as suspicious.

We do not believe right now that there is a simpler way to create this signature,
as the signer does not know the root of the `public_key` or `target_key` of `TxOut`'s
relative to the ristretto basepoint. The signer does know the root relative to some of
the public subaddress keys of the recipient, but such a signature cannot be verified

> by someone who does not know who the recipient.

Reviewer: ```suggestion
by someone who does not know who the recipient is.
```

- Feature Name: Standardized Sender Signature
- Start Date: 2022-03-14
- MCIP PR: [mobilecoinfoundation/mcips#0030](https://github.com/mobilecoinfoundation/mcips/pull/0030)
- MobileCoin Epic: None

# Summary
[summary]: #summary

> Enact MCIP #18 (Janus Mitigation), which sets the ephemeral key in the encrypted fog hint

Reviewer 3: We haven't merged MCIP 18, and it doesn't look like anyone has reviewed it yet. I think that needs to be reviewed and merged before we build on it. For example, I'm having a hard time finding the "fog hint ephemeral key" in the source, so it's hard for me to evaluate what's going on here.

Reviewer 2: For sure, we should definitely review that MCIP and merge it separately if we want to accept this MCIP, IMO

to be the tx private key times the Ristretto base point. The original motivation was to enact
Janus attack mitigation.

We find a secondary use-case to motivate this: it becomes possible for the signer to create

> a Schnorr signature over any payload, which a verifier can validate against the keys appearing

Reviewer 3: To make sure I'm following this: the public key that a sender would sign against is in the encrypted fog hint, so only the recipient's fog server can obtain this key and use it to verify messages signed with it?

Reviewer 2: so, the way that cryptobox works is:

* I get your public key (a Ristretto curve point)
* I choose a random scalar (private fog key, `r_fog` in Koe's notation in Mechanics of Mobilecoin)
* I do key exchange against your public key and `r_fog`, and I put `r_fog * G` as the first 32 bytes of the e_fog_hint.
* I use the key resulting from key exchange to encrypt and mac my payload
* I append 2 versioning bytes.

So the public key that a sender would sign against is `r_fog`, and can be validated against `r_fog * G`, which is the first 32 bytes of fog hint.

Koe proposed that instead of having two independent random keys `r` and `r_fog`, we should just set `r_fog = r`, and this enables a Janus mitigation test, without really impacting fog since the `r_fog` number is a perfectly fine secret key anyways.

So this breaks some encapsulation that's separating the `e_fog_hint` field from the rest of the Tx math. But, it serves several useful purposes, and it's not that big a deal from the point of cryptobox / fog.

However, these numbers are useless for convincing a third-party who is not a party to the transaction,
because only the recipient is able to validate these numbers. There is currently no form of receipt that

> can be validated by anyone, just from the blockchain.

Reviewer: Is this an observation about the current state of things, or an indication that this MCIP solves this problem? My understanding is that this MCIP doesn't propose to have on-chain receipts that can be validated by anyone, but that the off-chain store could be, is that correct?
